lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,376 @@
+
+        df = self.trace.analysis.tasks.df_task_activation(task)
+        df = df[['active', 'cpu']]
+
+        df_freq = self.trace.analysis.frequency.df_cpus_frequency()
+        df_freq = df_freq[['cpu', 'frequency']]
+        df_freq = df_freq.pivot(index=None, columns='cpu', values='frequency')
+        df_freq.reset_index(inplace=True)
+        df_freq.set_index('Time', inplace=True)
+
+        df = df.merge(df_freq, how='outer', left_index=True, right_index=True)
+
+        # Ensures that frequency values are propogated through the entire
+        # DataFrame, as it is possible that no frequency event occur
+        # during a phase.
+        for cpu in self.plat_info['cpu-capacities']['rtapp']:
+            df[cpu].ffill(inplace=True)
+
+        return df
+
+    def _get_phases_df(self):
+        task = self.rtapp_task_ids_map['task'][0]
+
+        df = self.trace.analysis.rta.df_phases(task).copy()
+        df.reset_index(inplace=True)
+        df.rename(columns={'index': 'start'}, inplace=True)
+        df['end'] = df['start'].shift(-1)
+
+        phases = [(0, 0, 0)] + self._get_phases(self.plat_info)
+        df['uclamp_val'] = df.apply(
+                lambda x: phases[int(x.phase)][0],
+                axis=1)
+
+        # TODO: To remove once we have named phases.
+        df = df[df.phase > 0]
+
+        return df
+
+    def _for_each_phase(self, callback):
+        df_phases = self._get_phases_df()
+        df_trace = self._get_trace_df()
+
+        def parse_phase(phase):
+            start = phase['start']
+            end = phase['end']
+            df = df_trace
+
+            # During a phase change, rt-app will wakeup and then change
+            # UtilClamp value will be changed. We then need to wait for the
+            # second wakeup for the kernel to apply the most recently set
+            # UtilClamp value.
+            start = df[(df.index >= start) &
+                       (df['active'] == 1)].first_valid_index()
+
+            end = end if not np.isnan(end) else df.last_valid_index()
+
+            if (start >= end):
+                raise ValueError('Phase ends before it has even started')
+
+            df = df_trace[start:end].copy()
+
+            return callback(df, phase)
+
+        return df_phases.apply(parse_phase, axis=1)
+
+    def _parse_phases(self, callback, metrics):
+        res = ResultBundle.from_bool(self._for_each_phase(callback).all())
+        res.add_metric('Phases', metrics)
+
+        return res
+
+    def _plot_phases(self, test, failures, signal=None):
+        task = self.rtapp_task_ids_map['task'][0]
+        ax = self.trace.analysis.tasks.plot_task_activation(task,
+                                                            which_cpu=True)
+        ax = self.trace.analysis.rta.plot_phases(task, axis=ax)
+        for failure in failures:
+            ax.axvline(failure, alpha=0.5, color='r')
+        if isinstance(signal, pd.DataFrame):
```

Comment:
The idiom is to check for None:

```
if signal is not None
```

Which is easier to spot and will avoid swallowing the input silently instead of:
* failing loudly with some exception if it gets passed wrongfully e.g. an int (we want exceptions in these cases, since there is no "compile-time" check for that)
* Simply working with duck-typing, e.g. with a Series since Series.plot() is essentially the same as DataFrame.plot()

Reply:
done. Can't recall why I did that ...